api/service/log_service.go

```diff
@@ -15,53 +15,290 @@
 )
 
+// LogLine represents a single log line from a container running in Kubernetes.
+type LogLine struct {
+	// Log timestamp in RFC3339 format
+	Timestamp time.Time `json:"timestamp"`
+	// Kubernetes namespace where the pod running the container is created
+	Namespace string `json:"namespace"`
+	// Pod name running the container that produces this log
+	PodName string `json:"pod_name"`
+	// Container name that produces this log
+	ContainerName string `json:"container_name"`
+	// Log in text format
+	TextPayload string `json:"text_payload,omitempty"`
+
+	PrefixColor *color.Color `json:"-"`
+}
+
+func (l LogLine) GenerateText(options LogQuery) string {
+	text := []string{}
+
+	if options.Prefix != "" {
+		p := l.PrefixColor.SprintFunc()
+		if options.Prefix == prefixAddPodContainer {
+			text = append(text, p(l.PodName)+" "+p(l.ContainerName))
+		} else if options.Prefix == prefixAddPod {
+			text = append(text, p(l.PodName))
+		} else if options.Prefix == prefixAddContainer {
+			text = append(text, p(l.ContainerName))
+		}
+	}
+
+	text = append(text, l.Timestamp.Format(time.RFC3339), l.TextPayload)
+	return strings.Join(text, " ") + "\n"
+}
+
+type ReadLogStream struct {
+	Stream        io.ReadCloser
+	PodName       string
+	ContainerName string
+}
+
 type LogService interface {
-	ReadLog(options *LogQuery) (io.ReadCloser, error)
+	StreamLogs(logLineCh chan string, stopCh chan struct{}, options *LogQuery) error
 }
 
 type logService struct {
 	// map of cluster name to cluster client
-	logClients map[string]corev1.CoreV1Interface
+	clusterControllers map[string]cluster.Controller
 }
 
 // NewLogService create a log service
-// logClients is a map of cluster name to its cluster client (CoreV1Interface)
-func NewLogService(logClients map[string]corev1.CoreV1Interface) LogService {
-	return &logService{logClients: logClients}
+// clusterControllers is a map of cluster name to its cluster.Controller
+func NewLogService(clusterControllers map[string]cluster.Controller) LogService {
+	color.NoColor = false // To bypasses the check for non-tty output streams.
+
+	return &logService{clusterControllers: clusterControllers}
 }
 
-// ReadLog read log from a container with the given log options
-// Return ReadCloser object which can be used to stream the log when the LogOptions.Follow is set to true
-func (l logService) ReadLog(options *LogQuery) (io.ReadCloser, error) {
-	logClient, ok := l.logClients[options.Cluster]
+func (l logService) StreamLogs(logLineCh chan string, stopCh chan struct{}, options *LogQuery) error {
+	clusterController, ok := l.clusterControllers[options.Cluster]
 	if !ok {
-		return nil, fmt.Errorf("unable to find cluster %s", options.Cluster)
+		return fmt.Errorf("unable to find cluster %s", options.Cluster)
+	}
+
+	namespace := options.Namespace
+	labelSelector := l.getLabelSelector(*options)
+
+	pods, err := clusterController.ListPods(namespace, labelSelector)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		for {
```

Comment:
I think it's more idiomatic in go to use ticker. E.g.

```
ticker := time.NewTicker(500 * time.Millisecond)
    go func() {
        for {
            select {
            case <-stopCh:
                return
            case t := <-ticker.C:
                doStuff
            }
        }
    }()
```

Reply:
Thanks for the suggestion. Updated!

Reply:
I think it can also further improved by breaking up the log collection from different pods.

```
ticker := time.NewTicker(1 * time.Second) 
go func() { 
for { 
   select { 
     case <-stopCh: return 
     case t := <-ticker.C: 
       allLogLines, err  :=  readPodLogs(pods) 
       	for _, logLine := range allLogLines {
		logLineCh <- logLine.generateText(*options)
	}
    } 
  } 
}()


```

Reply:
Done that. Please take a look again

Reply:
Great!